### PR TITLE
vector_search: fix SELECT on local vector index

### PR DESCRIPTION
**Pull request by knowack1**
Queries against local vector indexes were failing with the error:
```ANN ordering by vector requires the column to be indexed using 'vector_index'```

This was a regression introduced by https://github.com/scylladb/scylladb/commit/15788c373484094ee1a3464c89137f7807a3eaf9, which incorrectly
assumed the first column in the targets list is always the vector column.
For local vector indexes, the first column is the partition key, causing
the failure.

Previously, serialization logic for the target index option was shared
between vector and secondary indexes. This is no longer viable due to
the introduction of local vector indexes and vector indexes with filtering
columns, which have  different target format.

This commit introduces a dedicated JSON-based serialization format for
vector index targets, identifying the target column (tc), filtering
columns (fc), and partition key columns (pk). This ensures unambiguous
serialization and deserialization for all vector index types.

This change is backward compatible for regular vector indexes. However,
it breaks compatibility for local vector indexes and vector indexes with
filtering columns created in version 2026.1.0. To mitigate this, usage
of these specific index types will be blocked in the 2026.1.0 release
by failing ANN queries against them in vector-store service.

Fixes: SCYLLADB-895

Backport to 2026.1 is required as this issue occurs also on this branch.

### CODE REVIEW

File: index/vector_index.cc [hunk 1]
```diff
@@ -105,17 +105,38 @@ const static std::unordered_map<sstring, std::function<void(const sstring&, cons
...
+        }
+        return sstring(rjson::to_string_view(ck->GetArray()[0]));
+    }
+
+    // Global index: pk should contain only the vector column
+    if (pk->Size() != 1) {
+        throw exceptions::invalid_request_exception("Invalid target option: pk should contain only the vector column for a global index");
     }
-    return target_parser::get_target_column_name_from_string(targets);
+    return sstring(rjson::to_string_view(pk->GetArray()[0]));
```
[the remark this answers is not in the document]
Comment (knowack1) [reply index/vector_index.cc]: Done

File: index/vector_index.cc [hunk 2]
```diff
@@ -105,17 +105,38 @@ const static std::unordered_map<sstring, std::function<void(const sstring&, cons
...
 
-sstring get_vector_index_target_column(const sstring& targets) {
+
+static bool is_local_index(rjson::value* pk, const schema& schema) {
+    // A local index has its pk matching the table's partition key columns exactly.
+    auto pk_columns = schema.partition_key_columns();
+    auto pk_array = pk->GetArray();
+    return std::equal(pk_columns.begin(), pk_columns.end(), pk_array.begin(), pk_array.end(), [](const column_definition& col, const rjson::value& val) {
+        return col.name_as_text() == sstring(rjson::to_string_view(val));
+    });
```
[the remark this answers is not in the document]
Comment (knowack1) [reply index/vector_index.cc]: The check is done in upper function `get_vector_index_target_column`

File: index/vector_index.cc [hunk 3]
```diff
@@ -105,17 +105,43 @@ const static std::unordered_map<sstring, std::function<void(const sstring&, cons
...
+}
+
+static sstring get_vector_index_target_column(const schema& schema, const sstring& targets) {
     std::optional<rjson::value> json_value = rjson::try_parse(targets);
     if (!json_value || !json_value->IsObject()) {
-        return target_parser::get_target_column_name_from_string(targets);
+        // Global index with a single target column
+        return targets;
```
[the remark this answers is not in the document]
Comment (knowack1) [reply index/vector_index.cc]: Fixed

File: index/vector_index.cc [hunk 4]
```diff
@@ -105,17 +105,43 @@ const static std::unordered_map<sstring, std::function<void(const sstring&, cons
...
 
-sstring get_vector_index_target_column(const sstring& targets) {
+
+static bool is_local_index(rjson::value* pk, const schema& schema) {
+    // A local index has its pk matching the table's partition key columns exactly.
+    auto pk_columns = schema.partition_key_columns();
+    auto pk_array = pk->GetArray();
+    return std::equal(pk_columns.begin(), pk_columns.end(), pk_array.begin(), pk_array.end(), [](const column_definition& col, const rjson::value& val) {
+        return col.name_as_text() == sstring(rjson::to_string_view(val));
+    });
```
[the remark this answers is not in the document]
Comment (knowack1) [reply index/vector_index.cc]: The same logic to determina local vs global vector index is currently present is vector-store. Additionally logic to determine index locality for standard index does not match for vector index.

File: index/vector_index.cc [hunk 5]
```diff
@@ -105,17 +105,43 @@ const static std::unordered_map<sstring, std::function<void(const sstring&, cons
...
+        return cql3_parser::index_target::column_name_from_target_string(targets);
     }
 
     rjson::value* pk = rjson::find(*json_value, "pk");
-    if (pk && pk->IsArray() && !pk->Empty()) {
-        return sstring(rjson::to_string_view(pk->GetArray()[0]));
+    rjson::value* ck = rjson::find(*json_value, "ck");
+
+    if (!pk || !pk->IsArray() || pk->Empty()) {
+        throw exceptions::invalid_request_exception("Invalid vector index target option: 'pk' field is missing, not an array, or empty");
```
Comment (ewienik) [inline index/vector_index.cc]: I doubt it is not a request exception. 'pk' field seems an internal structure in index options. If this is a real request exception, then message would be more meaningful for the user.

File: index/vector_index.cc [hunk 6]
```diff
@@ -105,17 +105,43 @@ const static std::unordered_map<sstring, std::function<void(const sstring&, cons
...
+    }
+
+    if (is_local_index(pk, schema)) {
+        // Local index: target column is the first element of ck
+        if (!ck || !ck->IsArray() || ck->Empty()) {
+            throw exceptions::invalid_request_exception("Invalid vector index target option: 'ck' field is missing, not an array, or empty for local index");
```
Comment (ewienik) [inline index/vector_index.cc]: Similar case as with the previous request exception.

**Review by ewienik — APPROVED** [review]

File: index/vector_index.cc (hunk 5 above)
Comment (knowack1) [reply index/vector_index.cc]: Fixed.
You are right. This exception should never happen unless someone will modify targets serialization code without updating deserialization code accordingly. Hence this is rather internal error.  Changed to use on_internal_error.

File: index/vector_index.cc (hunk 6 above)
Comment (knowack1) [reply index/vector_index.cc]: Fixed.
Changed to on_internal_error.

File: test/vector_search/vector_store_client_test.cc [hunk 7]
```diff
@@ -1243,3 +1212,26 @@ SEASTAR_TEST_CASE(vector_store_client_vector_index_with_additional_filtering_col
...
+    auto server = co_await make_vs_mock_server();
+
+    auto cfg = make_config();
+    cfg.db_config->vector_store_primary_uri.set(format("http://server.node:{}", server->port()));
+    co_await do_with_cql_env(
+            [&](cql_test_env& env) -> future<> {
+                auto schema = co_await create_test_table(env, "ks", "test");
+                auto& vs = env.local_qp().vector_store_client();
+                configure(vs).with_dns({{"server.node", std::vector<std::string>{server->host()}}});
+                vs.start_background_tasks();
```
Comment (nuivall) [inline test/vector_search/vector_store_client_test.cc]: this boilerplate in each test probably could be put in common place in the future. E.g. `do_with_cql_env_with_vector_search`

File: index/vector_index.cc [hunk 8]
```diff
@@ -105,17 +106,43 @@ const static std::unordered_map<sstring, std::function<void(const sstring&, cons
...
     rjson::value* pk = rjson::find(*json_value, "pk");
-    if (pk && pk->IsArray() && !pk->Empty()) {
-        return sstring(rjson::to_string_view(pk->GetArray()[0]));
+    rjson::value* ck = rjson::find(*json_value, "ck");
+
+    if (!pk || !pk->IsArray() || pk->Empty()) {
+        utils::on_internal_error("Invalid vector index target option: 'pk' field is missing, not an array, or empty");
+    }
+
+    if (is_local_index(pk, schema)) {
```
Comment (nuivall) [inline index/vector_index.cc]: is_vector_index_on_column has index_metadata& im which has local(), can't this be used here?

File: test/vector_search/vector_store_client_test.cc (hunk 7 above)
Comment (knowack1) [reply test/vector_search/vector_store_client_test.cc]: Done.
Added new commit with common setup for some tests.

File: index/vector_index.cc (hunk 8 above)
Comment (knowack1) [reply index/vector_index.cc]: The `im.local()` method (wrongly) returns true for this type of index (which is a global vector index):
```
CREATE CUSTOM INDEX idx ON ks.test (embedding, ck1) USING 'vector_index'
```

However, there seems to be an inconsistency (possibly a bug) in the im.local() behavior.

When the index metadata is created for the first time in [create_index_statement.cc#L704](https://github.com/scylladb/scylladb/blob/master/cql3/statements/create_index_statement.cc#L704), `im.local()` is `false` (correct) for this particular case.

Later, however, after deserializing the "localness" from the target options in [secondary_index.cc#L80](https://github.com/scylladb/scylladb/blob/master/index/secondary_index.cc#L80), it becomes `true` (wrong).

This seems to be an issue with localness serialization/deserialization within the index metadata.

If possible, I would prefer not to touch this sensitive code within the scope of this PR. I recommend creating a dedicated issue for this problem to carefully rethink and prepare a fix (if we decide a fix is necessary). Is this acceptable?

Comment (nuivall) [reply index/vector_index.cc]: I'd say it'd be better to fix original local() function than adding separate is_local_index() one, it's creating unnecessary tech debt.

Comment (knowack1) [reply index/vector_index.cc]: Done.
Please see commit and its description: https://github.com/scylladb/scylladb/pull/28862/changes/b0fa5c22ceda39c93d64d869d8bdfada4f2f3d0f

**Review by ewienik — APPROVED** [review]

File: index/secondary_index.cc [hunk 9]
```diff
@@ -139,19 +139,26 @@ sstring target_parser::serialize_targets(const std::vector<::shared_ptr<cql3::st
...
-        rjson::push_back(pk_array, std::move(pk_json));
-        pk_json = std::move(pk_array);
-    }
-    rjson::add_with_string_name(json_map, PK_TARGET_KEY, std::move(pk_json));
-    if (targets.size() > 1) {
+
+    bool local = std::holds_alternative<index_target::multiple_columns>(targets.front()->value);
+    if (local) {
+        // Local index: pk = base table partition key columns,
+        //              ck = remaining target columns.
```
Comment (nyh) [inline index/secondary_index.cc]: I think we've made a mistake mixing **local vector indexes** and **local secondary indexes**. These are very different things, and even if the CQL syntax for both is similar, they mean completely different things... In secondary index, these columns are **key** columns. In vector index, they are **projected** (filtering-only) columns. there was no reason why we needed to encode them identically in the "targets" field.

In particular, here in this comment, I don't understand - does it really apply to a "local index", i.e., the classic local secondary index, or is it a hack for local **vector** index? You mention "ck", but are these really clustering keys or what?

File: index/secondary_index.cc [hunk 10]
```diff
@@ -139,19 +139,26 @@ sstring target_parser::serialize_targets(const std::vector<::shared_ptr<cql3::st
...
-    rjson::value pk_json = std::visit(as_json_visitor(), targets.front()->value);
-    if (!pk_json.IsArray()) {
-        rjson::value pk_array = rjson::empty_array();
-        rjson::push_back(pk_array, std::move(pk_json));
-        pk_json = std::move(pk_array);
-    }
-    rjson::add_with_string_name(json_map, PK_TARGET_KEY, std::move(pk_json));
-    if (targets.size() > 1) {
+
+    bool local = std::holds_alternative<index_target::multiple_columns>(targets.front()->value);
```
Comment (nyh) [inline index/secondary_index.cc]: I'm having a hard time figuring out what is really different about this new code. Is this holds_alternative<> thing the same as targets.size() > 1 we had in the old code, or not the same?

**Review by nyh — CHANGES_REQUESTED** [review]
You claimed that there is no change to the encoding of the "targets" of supported index types (classic secondary index, local secondary index, and global vector index).  But I don't see any test-based evidence for this. How did you actually test this? It won't be tested on traditional test/cqlpy and test/alternator tests, because we'll still use the same matching target generation and parsing code. I am guessing we don't have "upgrade" tests to verify that the targets written by the new code are identical to the old code. 

So I think it will be a good idea to write a unit test, can even be written in test/cqlpy or in C++ or test/cql, which creates an index of some type (e.g., a classic secondary index), and then reads the "target" from the schema table and checks it is equal to something. After your patch it should remain unchanged. Such a test would convince me that your patch doesn't cause a regression.

**Comment by nuivall** [comment]
> > @nuivall I am not sure if there are any automatic upgrade test cases, but I manually verified that GSIs and LSIs created before the change still work with these commits - meaning the change is backward compatible.
> 
> I also don't know if there are upgrade tests, so I assume there are none :-( See my suggestion above for a very simple regression test, can be 5 lines of test/cqlpy lines code for each scenario - create a secondary index, read the "targets" from the schema, compare it to a known value (the test should pass before and after this patch).

there is `version=scylla_2025_1` fixture in cluster tests

File: index/vector_index.cc [hunk 11]
```diff
@@ -105,15 +105,70 @@ const static std::unordered_map<sstring, std::function<void(const sstring&, cons
...
+    }
+
+    return rjson::print(json_map);
+}
+
 sstring get_vector_index_target_column(const sstring& targets) {
     std::optional<rjson::value> json_value = rjson::try_parse(targets);
     if (!json_value || !json_value->IsObject()) {
         return target_parser::get_target_column_name_from_string(targets);
     }
```
[the remark this answers is not in the document]
Comment (knowack1) [reply index/vector_index.cc]: Fixed

File: cql3/statements/create_index_statement.cc [hunk 12]
```diff
@@ -697,7 +698,11 @@ index_metadata create_index_statement::make_index_metadata(const std::vector<::s
                                                            const index_options_map& options)
 {
     index_options_map new_options = options;
-    auto target_option = secondary_index::target_parser::serialize_targets(targets);
+    auto class_it = options.find(db::index::secondary_index::custom_class_option_name);
+    bool is_vector = class_it != options.end() && class_it->second == "vector_index";
+    auto target_option = is_vector
+        ? secondary_index::vector_index::serialize_targets(targets)
+        : secondary_index::target_parser::serialize_targets(targets);
```
[the remark this answers is not in the document]
Comment (knowack1) [reply cql3/statements/create_index_statement.cc]: Fixed

File: test/cqlpy/test_vector_index.py [hunk 13]
```diff
@@ -9,6 +9,7 @@
 import pytest
 from .util import new_test_table, is_scylla, unique_name
 from cassandra.protocol import InvalidRequest, ConfigurationException
+import json
```
[the remark this answers is not in the document]
Comment (knowack1) [reply test/cqlpy/test_vector_index.py:13]: Fixed

File: test/cqlpy/test_secondary_index.py [hunk 14]
```diff
@@ -2031,6 +2032,52 @@ def test_index_in_system_schema_indexes(cql, built_index):
     assert res[0].kind == 'COMPOSITES'
     assert res[0].options == {'target': 'v'}
 
+# Test that the "target" option in system_schema.indexes is serialized
+# correctly for global secondary indexes on collection columns.
+# This format is critical for backward compatibility, as it's read from
+# disk on startup to rebuild indexes. An incompatible change would prevent
+# existing local indexes from being recreated after an upgrade.
```
[the remark this answers is not in the document]
Comment (knowack1) [reply test/cqlpy/test_secondary_index.py]: Fixed

File: index/vector_index.cc [hunk 15]
```diff
@@ -105,15 +105,70 @@ const static std::unordered_map<sstring, std::function<void(const sstring&, cons
...
+        return index_target::escape_target_column(*std::get<index_target::single_column>(tc));
+    }
+
+    const bool has_pk = std::holds_alternative<index_target::multiple_columns>(targets.front()->value);
+    const size_t tc_idx = has_pk ? 1 : 0;
+    const size_t fc_count = targets.size() - tc_idx - 1;
+
+    rjson::value json_map = rjson::empty_object();
+    rjson::add_with_string_name(json_map, TC_TARGET_KEY, rjson::from_string(std::get<index_target::single_column>(targets[tc_idx]->value)->text()));
+
```
[the remark this answers is not in the document]
Comment (knowack1) [reply index/vector_index.cc:189]: Fixed

File: test/cqlpy/test_vector_index.py [hunk 16]
```diff
@@ -264,6 +265,64 @@ def test_vector_index_version_fail_given_as_option(cql, test_keyspace, scylla_on
...
+# option to identify the target column.
+def test_vector_index_target_serialization(cql, test_keyspace, scylla_only, skip_without_tablets):
+    schema = 'p int primary key, v vector<float, 3>'
+    with new_test_table(cql, test_keyspace, schema) as table:
+        index_name = unique_name()
+        cql.execute(f"CREATE CUSTOM INDEX {index_name} ON {table}(v) USING 'vector_index'")
+
+        res = [r for r in cql.execute('select * from system_schema.indexes')
+               if r.index_name == index_name]
+
```
[the remark this answers is not in the document]
Comment (knowack1) [reply test/cqlpy/test_vector_index.py:313]: This is according to what is currently in test cases.

File: test/boost/vector_index_test.cc [hunk 17]
```diff
@@ -0,0 +1,189 @@
...
+            make_single("MyCol"),
+            make_single("f1"),
+    };
+    const auto serialized = vector_index::serialize_targets(targets);
+
+    BOOST_CHECK_EQUAL(vector_index::get_target_column(serialized), "MyCol");
+}
+
+BOOST_AUTO_TEST_CASE(get_target_column_reutrns_column_with_quotes_from_json) {
+    std::vector<::shared_ptr<index_target>> targets = {
```
[the remark this answers is not in the document]
Comment (knowack1) [reply test/boost/vector_index_test.cc]: Fixed

File: index/vector_index.cc [hunk 18]
```diff
@@ -105,17 +105,83 @@ const static std::unordered_map<sstring, std::function<void(const sstring&, cons
...
+        for (size_t i = tc_idx + 1; i < targets.size(); ++i) {
+            if (!std::holds_alternative<index_target::single_column>(targets[i]->value)) {
+                throw exceptions::invalid_request_exception("Vector index filtering column must be a single column");
+            }
+            rjson::push_back(fc_json, rjson::from_string(std::get<index_target::single_column>(targets[i]->value)->text()));
+        }
+        rjson::add_with_string_name(json_map, FC_TARGET_KEY, std::move(fc_json));
+    }
+
+    return rjson::print(json_map);
```
[the remark this answers is not in the document]
Comment (knowack1) [reply index/vector_index.cc:209]: Fixed

File: index/secondary_index.cc (hunk 9 above)
Comment (knowack1) [reply index/secondary_index.cc]: I think you are right. We were trying to use the existing target option logic from secondary indexes, which doesn't fit vector indexes because they have extra filtering columns that the secondary index implementation wasn't designed for. As a consequence, using the old logic leads to cryptic JSON for vector indexes, where interpretation requires "hidden knowledge."

Please review my recent updates to this PR:
- Introduced pytest cases for secondary index target option serialization to prevent backward-incompatibility.
- Introduced pytest cases for regular vector index target option serialization to prevent backward-incompatibility.
- Introduced a new JSON schema for local vector indexes and vector indexes with filtering columns:
```
//   (v)                          -> "v"
//   (v, f1, f2)                  -> {"tc":"v","fc":["f1","f2"]}
//   ((p1, p2), v)                -> {"tc":"v","pk":["p1","p2"]}
//   ((p1, p2), v, f1, f2)        -> {"tc":"v","pk":["p1","p2"],"fc":["f1","f2"]}
```
- Fixed DESC INDEX for local vector indexes and vector indexes with filtering columns (this is an issue that already exists on master).

File: test/cqlpy/test_secondary_index.py [hunk 19]
```diff
@@ -2031,6 +2032,51 @@ def test_index_in_system_schema_indexes(cql, built_index):
...
+            assert res[0].options == {'target': expected_target}, \
+                f"For index expression '{index_expr}': expected target '{expected_target}', got '{res[0].options}'"
+
+# Test that the "target" option in system_schema.indexes is serialized
+# correctly for local secondary indexes. This format is critical for
+# backward compatibility, as it's read from disk on startup to rebuild
+# indexes. An incompatible change would prevent existing local indexes
+# from being recreated after an upgrade.
+def test_local_index_target_serialization(cql, test_keyspace, scylla_only):
+    schema = "a int, b int, c int, v int, PRIMARY KEY ((a, b), c)"
```
Comment (nuivall) [inline test/cqlpy/test_secondary_index.py:2098]: why testing index ON different column types than global index? only on `v int` while global included collections

File: test/cqlpy/test_vector_index.py [hunk 20]
```diff
@@ -283,6 +284,45 @@ def test_vector_index_target_serialization(cql, test_keyspace, scylla_only, skip
         assert len(res) == 1
         assert res[0].options['target'] == 'v'
 
+# Test "target" option serialization for vector index with filtering columns.
+def test_vector_index_target_serialization_filtering_columns(cql, test_keyspace, scylla_only, skip_without_tablets):
```
Comment (nuivall) [inline test/cqlpy/test_vector_index.py:318]: why not adding those tests together with in [index: test: vector index target option serialization test](https://github.com/scylladb/scylladb/pull/28862/changes/fff1fe06ed08315ef5ca3dda8dc18b5e412db929) ?

**Review by nuivall — APPROVED** [review]

File: test/cqlpy/test_vector_index.py (hunk 20 above)
Comment (knowack1) [reply test/cqlpy/test_vector_index.py:318]: These tests will not pass in this commit because they validate changes introduced here. While the tests in mentioned commit are intended to test current behavior on the master branch (which was previously untested) to ensure no regressions occur.

File: test/cqlpy/test_secondary_index.py (hunk 19 above)
Comment (knowack1) [reply test/cqlpy/test_secondary_index.py:2098]: I added collection-related tests to basically cover this if statement: https://github.com/scylladb/scylladb/blob/854c374ebf2b6a309d786ae510108c2de1b618af/index/secondary_index.cc#L110, which has special treatment for collections. Since this special treatment relates only to global secondary indexes, I did not cover the same for local indexes.

File: test/cqlpy/test_secondary_index.py [hunk 21]
```diff
@@ -2031,6 +2032,51 @@ def test_index_in_system_schema_indexes(cql, built_index):
     assert res[0].kind == 'COMPOSITES'
     assert res[0].options == {'target': 'v'}
 
+# Test that the "target" option in system_schema.indexes is serialized
+# correctly for secondary indexes on collection columns.
+# This format is critical for backward compatibility, as it's read from
+# disk on startup to rebuild indexes. An incompatible change would prevent
+# existing indexes from being recreated after an upgrade.
+def test_global_collection_index_target_serialization(cql, test_keyspace):
```
Comment (nyh) [inline test/cqlpy/test_secondary_index.py:2040]: This test also passes on Cassandra? Nice. We are compatible in system_schema.indexes for all these strange index types? Please check with test/cqlpy/run-cassandra. Thanks.
(of course if the answer is no, this test should be marked scylla_only).

File: test/cqlpy/test_secondary_index.py (hunk 19 above)
Comment (nyh) [reply test/cqlpy/test_secondary_index.py:2098]: Indeed. Local secondary index on collections is not yet supported (https://github.com/scylladb/scylladb/issues/10713). Maybe one day it will, but nobody is even thinking of working on it.

File: test/cqlpy/test_secondary_index.py [hunk 22]
```diff
@@ -2031,6 +2032,51 @@ def test_index_in_system_schema_indexes(cql, built_index):
...
+    with new_test_table(cql, test_keyspace, schema) as table:
+        index_name = unique_name()
+        cql.execute(f"CREATE INDEX {index_name} ON {table}((a, b), v)")
+        wait_for_index(cql, test_keyspace, index_name)
+
+        res = [r for r in cql.execute('select * from system_schema.indexes')
+               if r.index_name == index_name]
+
+        assert len(res) == 1
+        assert json.loads(res[0].options['target']) == {"pk": ["a", "b"], "ck": ["v"]}
```
Comment (nyh) [inline test/cqlpy/test_secondary_index.py:2109]: By the way, an implied assertion being made here is that the LSI target serialization is in json form, maybe worth adding a comment saying that.

File: test/cqlpy/test_secondary_index.py [hunk 23]
```diff
@@ -2031,6 +2032,51 @@ def test_index_in_system_schema_indexes(cql, built_index):
...
+            assert len(res) == 1
+            assert res[0].kind == 'COMPOSITES'
+            assert res[0].options == {'target': expected_target}, \
+                f"For index expression '{index_expr}': expected target '{expected_target}', got '{res[0].options}'"
+
+# Test that the "target" option in system_schema.indexes is serialized
+# correctly for local secondary indexes. This format is critical for
+# backward compatibility, as it's read from disk on startup to rebuild
+# indexes. An incompatible change would prevent existing local indexes
+# from being recreated after an upgrade.
```
Comment (nyh) [inline test/cqlpy/test_secondary_index.py:2096]: By the way, check out commit 2c244c6e09c4693a47eadded930651599f8145d1 which fixed a bug in the target serialization when the column name had special characters like braces, and explains how the target serialization works in different cases - it might be useful to add tests for the target encoding in those cases as well.

Also please consider updating docs/dev/secondary_index.md which only has an incomplete explanation of the format of "target", and your patch makes the description even less complete. I'm sure Claude can help you expand this documentation.

File: cql3/statements/create_index_statement.cc [hunk 24]
```diff
@@ -697,7 +699,11 @@ index_metadata create_index_statement::make_index_metadata(const std::vector<::s
                                                            const index_options_map& options)
 {
     index_options_map new_options = options;
-    auto target_option = secondary_index::target_parser::serialize_targets(targets);
+    auto class_it = options.find(db::index::secondary_index::custom_class_option_name);
+    bool is_vector = class_it != options.end() && boost::iequals(class_it->second, "vector_index");
```
Comment (nyh) [inline cql3/statements/create_index_statement.cc]: Nitpick (not important now): I think we should really have a boolean function to check "is it a vector index". What if in the future we'll want to allow a synonym for vector_index (I think we wanted to allow "sai"), or maybe a long Java-like name like is common in Cassandra?

File: index/vector_index.cc [hunk 25]
```diff
@@ -105,17 +105,79 @@ const static std::unordered_map<sstring, std::function<void(const sstring&, cons
...
+// Serialize vector index targets into a format using:
+//   "tc" for the target (vector) column,
+//   "pk" for partition key columns (local index),
+//   "fc" for filtering columns.
+// For a simple single-column vector index, returns just the column name.
+// Examples:
+//   (v)                          -> "v"
+//   (v, f1, f2)                  -> {"tc":"v","fc":["f1","f2"]}
+//   ((p1, p2), v)                -> {"tc":"v","pk":["p1","p2"]}
+//   ((p1, p2), v, f1, f2)        -> {"tc":"v","pk":["p1","p2"],"fc":["f1","f2"]}
```
Comment (nyh) [inline index/vector_index.cc:163]: By the way, only now it dawned on me that since we know the "class" of the index (regular secondary index or vector index), it doesn't matter at all if both types of indexes use the same encoding for "target" or not - since we can have different codes that encodes each type, and different code to decode each type. Right?

**Review by nyh — APPROVED** [review]
Looks good. I only had some minor comments for your consideration (in any case, we're in a merge freeze so can't merge this right away)

File: test/cqlpy/test_secondary_index.py (hunk 21 above)
Comment (knowack1) [reply test/cqlpy/test_secondary_index.py:2040]: No -  it fails. Cassandra stores `full(fl)` as the target for a frozen full collection index, but Scylla normalizes it to just the column name `fl`. The other collection types (`keys`, `values`, `entries`) are compatible.

Fixed by this if: https://github.com/scylladb/scylladb/pull/28862/changes/4dc28dfa52c287876063b8525f4c1f6ac836fe35#diff-272958765df2022b446f93523f762cd371f4bf18667929f43f7b1fb28f950b7eR2045

File: test/cqlpy/test_secondary_index.py (hunk 22 above)
Comment (knowack1) [reply test/cqlpy/test_secondary_index.py:2109]: Fixed. Comment added: https://github.com/scylladb/scylladb/pull/28862/changes/4dc28dfa52c287876063b8525f4c1f6ac836fe35#diff-272958765df2022b446f93523f762cd371f4bf18667929f43f7b1fb28f950b7eR2108

File: test/cqlpy/test_secondary_index.py (hunk 23 above)
Comment (knowack1) [reply test/cqlpy/test_secondary_index.py:2096]: Fixed.
Doc update + added test case `test_global_index_target_serialization_quoted_names`. Also added vector index dedicated dev doc: `docs/dev/vector_index.md`

File: cql3/statements/create_index_statement.cc (hunk 24 above)
Comment (knowack1) [reply cql3/statements/create_index_statement.cc]: Fixed. Added `is_vector_index`.

File: index/vector_index.cc (hunk 25 above)
Comment (knowack1) [reply index/vector_index.cc:163]: Exactly. Previously, we used shared serialization and deserialization logic for both vector and secondary indexes. However, that approach broke down after we introduced local vector indexes and vector indexes with filtering columns.

In the first iteration, I tried  keeping the same serialization but using dedicated deserialization. It felt a bit too much like 'magic' and resulted in confusing JSON; we were using pk and ck labels for what were actually vector and filtering columns.

This patch introduces fully dedicated serialization and deserialization paths for vector indexes. The JSON format is now specifically tailored for them, which makes the whole solution much cleaner and more intuitive.
